test/e2e/storage/host_path_type.go

```diff
@@ -0,0 +1,232 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"fmt"
+	"path"
+	"strings"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = utils.SIGDescribe("HostPathType", func() {
+	f := framework.NewDefaultFramework("host-path-type")
+
+	var (
+		ns           string
+		hostBaseDir  string
+		mountBaseDir string
+		basePod      *v1.Pod
+		config       *localTestConfig
+
+		hostPathUnset             v1.HostPathType
+		hostPathDirectoryOrCreate v1.HostPathType
+		hostPathDirectory         v1.HostPathType
+		hostPathFileOrCreate      v1.HostPathType
+		hostPathFile              v1.HostPathType
+		hostPathSocket            v1.HostPathType
+		hostPathCharDev           v1.HostPathType
+		hostPathBlockDev          v1.HostPathType
+	)
+
+	BeforeEach(func() {
+		ns = f.Namespace.Name
+
+		hostPathUnset = v1.HostPathUnset
+		hostPathDirectoryOrCreate = v1.HostPathDirectoryOrCreate
+		hostPathDirectory = v1.HostPathDirectory
+		hostPathFileOrCreate = v1.HostPathFileOrCreate
+		hostPathFile = v1.HostPathFile
+		hostPathSocket = v1.HostPathSocket
+		hostPathCharDev = v1.HostPathCharDev
+		hostPathBlockDev = v1.HostPathBlockDev
+
+		config = &localTestConfig{
+			ns:     f.Namespace.Name,
+			client: f.ClientSet,
+		}
+
+		By("Create a pod for further testing")
+		hostDir := "/var/tmp/hostpath-test"
+		mountDir := "/mnt/test"
+		basePod = f.PodClient().CreateSync(newHostPathTypeTestPod("", hostDir, mountDir, &hostPathDirectoryOrCreate))
+
+		By("Create a temporary clean directory for further testing")
+		result, err := utils.PodExec(basePod, fmt.Sprintf("mktemp -d -p %s hostpath.XXXXXX", mountDir))
+		Expect(err).NotTo(HaveOccurred())
+
+		mountBaseDir = result
+		hostBaseDir = path.Join(hostDir, strings.TrimPrefix(result, mountDir))
+	})
+
+	It("HostPath Directory Test", func() {
+		targetDir := path.Join(hostBaseDir, "adir")
+
+		By("Should fail on mounting non-existent directory 'adir' when HostPathType is HostPathDirectory")
```

Comment:
Do you want to make each of these cases it's own test case (it's own It())?  Then the outer It() could actually be Context() instead.